src/Everscale/Cell.cpp

```diff
@@ -0,0 +1,387 @@
+
+    explicit Reader(const uint8_t* _Nonnull buffer, size_t len) noexcept
+        : buffer(buffer), bufferLen(len) {
+    }
+
+    void require(size_t bytes) const {
+        if (offset + bytes > bufferLen) {
+            throw std::runtime_error("unexpected eof");
+        }
+    }
+
+    void advance(size_t bytes) {
+        offset += bytes;
+    }
+
+    inline const uint8_t* _Nonnull data() const {
+        return buffer + offset;
+    }
+
+    inline uint32_t readNextU32() {
+        const auto* _Nonnull p = data();
+        advance(sizeof(uint32_t));
+        return decode32BE(p);
+    }
+
+    inline size_t readNextUint(uint8_t len) {
+        const auto* _Nonnull p = data();
+        advance(len);
+
+        switch (len) {
+        case 1:
+            return static_cast<size_t>(*p);
+        case 2:
+            return static_cast<size_t>(decode16BE(p));
+        case 3:
+            return static_cast<size_t>(p[2]) | (static_cast<size_t>(p[1]) << 8) | (static_cast<size_t>(p[0]) << 16);
+        case 4:
+            return static_cast<size_t>(decode32BE(p));
+        default:
+            // Unreachable in valid cells
+            return 0;
+        }
+    }
+};
+
+std::shared_ptr<Cell> Cell::deserialize(const uint8_t* _Nonnull data, size_t len) {
+    Reader reader(data, len);
+
+    // Parse header
+    reader.require(6);
+    // 1. Magic
+    if (reader.readNextU32() != BOC_MAGIC) {
+        throw std::runtime_error("unknown magic");
+    }
+    // 2. Flags
+    struct Flags {
+        uint8_t refSize : 3;
+        uint8_t : 2; // unused
+        bool hasCacheBits : 1;
+        bool hasCrc : 1;
+        bool indexIncluded : 1;
+    };
+
+    static_assert(sizeof(Flags) == 1, "flags must be represented as 1 byte");
+    const auto flags = reinterpret_cast<const Flags*>(reader.data())[0];
+    const auto refSize = flags.refSize;
+    const auto offsetSize = reader.data()[1];
+    reader.advance(2);
+
+    // 3. Counters and root index
+    reader.require(refSize * 3 + offsetSize + refSize);
+    const auto cellCount = reader.readNextUint(refSize);
+    const auto rootCount = reader.readNextUint(refSize);
+    if (rootCount != 1) {
+        throw std::runtime_error("unsupported root count");
+    }
+    if (rootCount > cellCount) {
+        throw std::runtime_error("root count is greater than cell count");
+    }
+    const auto absent_count = reader.readNextUint(refSize);
+    if (absent_count > 0) {
+        throw std::runtime_error("absent cells are not supported");
+    }
+
+    reader.readNextUint(offsetSize); // total cell size
+
+    const auto rootIndex = reader.readNextUint(refSize);
+
+    // 4. Cell offsets (skip if specified)
+    if (flags.indexIncluded) {
+        reader.advance(cellCount * offsetSize);
+    }
+
+    // 5. Deserialize cells
+    struct IntermediateCell {
+        uint16_t bitLen;
+        Data data;
+        std::vector<size_t> references;
+    };
+
+    std::map<size_t, IntermediateCell> intermediate{};
```

Comment:
Does the map require to be ordered? Can we use `std::unordered_map` instead?

Reply:
Done